src/ptvsd/_vendored/pydevd/pydevd.py

```diff
@@ -2070,6 +2070,33 @@ def wait_for_commands(self, globals):
             time.sleep(0.01)
 
 
+class INetCommandsListener(object):
+
+    def before_send(self, net_command):
+        '''
+        Called just before a net command is sent to the IDE.
+
+        :type net_command: _pydevd_bundle.pydevd_net_command._BaseNetCommand
```

Comment:
This creates a dependency on `NetCommand`, and any changes to that would break the code.

At least for JSON messages, it feels like it would be better to pass `as_dict` directly instead. It has everything already, including message metadata, and it's just a regular `dict` instance, so the only thing one needs to interpret it is the DAP spec.